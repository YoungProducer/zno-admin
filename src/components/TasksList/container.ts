// Created by: Oleksandr Bezrukov
// Creation date: 4 February 2020

/**
 * Define main types and interfaces.
 * Define functions to get actions and varibales from redux store.
 * Export function which connect actions and store to 'TasksList' component.
 */

// External imports
import { connect } from 'react-redux';

// Application's imports
import { selectTasksList, selectTasksListOpen } from 'store/selectors/createTest';
import { deleteTaskAction, toggleOpenTasksListAction } from 'store/slices/createTest';
import { ITask } from 'store/slices/createTest';
import { RootState } from 'store/slices';

/**
 * Props which component get from the parent.
 */
interface IOwnProps {}

/**
 * Props which component get from the redux store.
 */
interface IStateProps {
    tasksList: ITask[];
    open: boolean;
}

/**
 * Props(actions) which component can dispatch to store
 */
interface IDispatchProps {
    deleteTask: (id: number) => void;
    toggleOpenTasksList: (payload: boolean) => void;
}

/**
 * Define type which describe all props which component get from the parent or from the store.
 */
export type TTasksListProps = IOwnProps & IStateProps & IDispatchProps;

/**
 * Define mapStateToProps function.
 * This function will connect variables from redux store to the component.
 */
const mapStateToProps = (state: RootState): IStateProps => ({
    tasksList: selectTasksList(state),
    open: selectTasksListOpen(state),
});

/**
 * Define mapDispatchToProps function.
 * This function will connect actions to the component.
 */
const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    deleteTask: (id: number) =>
        dispatch(deleteTaskAction(id)),

    toggleOpenTasksList: (payload: boolean) =>
        dispatch(toggleOpenTasksListAction(payload)),
});

/**
 * Export function which connect actions and variables from the redux store to component.
 */
export default connect<IStateProps, IDispatchProps, IOwnProps>(
    mapStateToProps,
    mapDispatchToProps,
);
