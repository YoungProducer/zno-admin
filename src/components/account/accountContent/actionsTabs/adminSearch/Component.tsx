import React, { useState, useEffect, useCallback, useRef } from 'react';
import TextField from '@material-ui/core/TextField';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import { Box, Button, Grow, Typography, Divider, MenuItem, Fade, CircularProgress, FormControl, InputLabel, InputAdornment } from '@material-ui/core';
import { Theme, createStyles, makeStyles } from '@material-ui/core/styles';

import _ from 'lodash';

import { SetSearchUserResultCredentials } from '../../../../../types/store/actionsCreators/users/searchResult';
import { Role } from '../../../../../types/store/actionsCreators/mainTypes';
import { FetchUpdateUserRootsCredentials } from '../../../../../types/store/actionsCreators/update/userRoot';

interface AdminSearchProps {
    foundUsers: SetSearchUserResultCredentials[];
    updateUserRootsLoading: boolean;
    updateUserRootsSuccess: boolean;
    findUserByEmailLoading: boolean;
    fetchFindUsersByEmail(credentials: string): any;
    fetchUpdateUserRoots(credentials: FetchUpdateUserRootsCredentials): any;
}

interface SelectedUser {
    id: string;
    email: string;
}

const useStyles = makeStyles((theme: Theme) => createStyles({
    list: {
        width: '100%',
        maxHeight: 300,
        overflow: 'auto',
    },
    divider: {
        marginTop: 10,
        marginBottom: 10,
    },
    button: {
        width: 240,
    },
    textField: {
        marginTop: 10,
        marginBottom: 10,
    },
    searchWrapper: {
        width: 240,
    },
    circularProgress: {
        marginLeft: 10,
    },
}));

const AdminSearch = ({
    foundUsers,
    updateUserRootsLoading,
    updateUserRootsSuccess,
    findUserByEmailLoading,
    fetchFindUsersByEmail,
    fetchUpdateUserRoots,
}: AdminSearchProps) => {
    const classes = useStyles();

    const [textFieldsValue, setTextFieldValue] = useState<string>('');
    const [search, toggleSearch] = useState<boolean>(false);
    const [typing, toggleTyping] = useState<boolean>(false);

    const roles: Role[] = [
        'ADMIN',
        'DEFAULT_USER',
    ];

    const [selectedRole, setSelectedRole] = useState<string>('DEFAULT_USER');
    const [selectedUser, setSelectedUser] = useState<SelectedUser>({ id: '', email: '' });

    const debounced = useRef(_.debounce((value: string) => {
        fetchFindUsersByEmail(value);
        toggleTyping(false);
    },                                  250));

    useEffect(() => debounced.current(textFieldsValue), [textFieldsValue]);
    useEffect(() => toggleTyping(true), [textFieldsValue]);

    return (
        <Box>
            <Typography
                variant="h4"
                color="primary"
                align="center"
            >
                Admin panel
            </Typography>
            <Divider className={classes.divider}/>
            <Button
                className={classes.button}
                color="secondary"
                onClick={() => toggleSearch(!search)}
            >
                {`${!search ? 'Give the roots' : 'Close'}`}
            </Button>
            <Grow
                in={search}
                unmountOnExit
                mountOnEnter
            >
                <div className={classes.searchWrapper}>
                    <TextField
                        className={classes.textField}
                        color="secondary"
                        value={textFieldsValue}
                        onChange={event => {
                            setTextFieldValue(event.target.value);
                        }}
                        InputProps={{
                            endAdornment: (
                                <InputAdornment position="end">
                                    { (findUserByEmailLoading || typing) &&
                                        <CircularProgress
                                            size="15px"
                                            color="secondary"
                                        />
                                    }
                                </InputAdornment>
                            ),
                        }}
                    />
                    <List className={classes.list}>
                        {foundUsers.map((user: SetSearchUserResultCredentials) => (
                            <ListItem
                                key={user.id}
                                button
                                onClick={() => setSelectedUser({ email: user.email, id: user.id })}
                            >
                                {user.email}
                            </ListItem>
                        ))}
                    </List>
                    <Fade
                        in={selectedUser.id !== ''}
                        mountOnEnter
                        unmountOnExit
                    >
                        <>
                            <Divider className={classes.divider}/>
                            <Typography variant="h5" color="secondary">
                                Selected user:
                            </Typography>
                            <Typography variant="h5">
                                { selectedUser.email }
                            </Typography>
                            <Divider className={classes.divider}/>
                            <TextField
                                className={classes.textField}
                                select
                                color="secondary"
                                value={selectedRole}
                                onChange={(event) => setSelectedRole(event.target.value)}
                            >
                                { roles.map((role) => (
                                    <MenuItem key={role} value={role}>
                                        { role }
                                    </MenuItem>
                                ))}
                            </TextField>
                            <Button
                                disabled={updateUserRootsLoading}
                                className={classes.button}
                                variant="outlined"
                                color="secondary"
                                onClick={() => fetchUpdateUserRoots({ userId: selectedUser.id, newRole: selectedRole })}
                            >
                                Update roots
                                { updateUserRootsLoading && <CircularProgress color="secondary" className={classes.circularProgress} size="30px" />}
                            </Button>
                        </>
                    </Fade>
                </div>
            </Grow>
        </Box>
    );
};

export default AdminSearch;
