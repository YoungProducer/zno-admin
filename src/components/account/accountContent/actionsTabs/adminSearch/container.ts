import { connect } from 'react-redux';
import { getFoundUsers } from '../../../../../store/selectors/users/searchResult';
import { createFetchFindUserByEmail } from '../../../../../store/actionsCreators/users/findByEmail';
import { FetchUpdateUserRootsCredentials } from '../../../../../types/store/actionsCreators/update/userRoot';
import { createFetchUpdateUserRoots } from '../../../../../store/actionsCreators/update/userRoots';
import { getUpdateUserRootsLoading, getUpdateUserRootsSuccess } from '../../../../../store/selectors/update/userRoots';
import { getFindUserByEmailLoading } from '../../../../../store/selectors/users/findByEmail';

const mapStateToProps = (state: any) => ({
    foundUsers: getFoundUsers(state.usersSearchResult),
    updateUserRootsLoading: getUpdateUserRootsLoading(state.updateUserRoots),
    updateUserRootsSuccess: getUpdateUserRootsSuccess(state.updateUserRoots),
    findUserByEmailLoading: getFindUserByEmailLoading(state.findUserByEmail),
});

const mapDispatchToProps = (dispatch: any) => ({
    fetchFindUsersByEmail: (email: string) => dispatch(createFetchFindUserByEmail(email)),
    fetchUpdateUserRoots: (credentials: FetchUpdateUserRootsCredentials) => dispatch(createFetchUpdateUserRoots(credentials)),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps,
);
