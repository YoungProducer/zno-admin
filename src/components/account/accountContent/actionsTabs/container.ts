import { connect } from 'react-redux';
import { getUserData } from '../../../../store/selectors/auth/signin';

const mapStateToProps = (state: any) => ({
    user: getUserData(state.signin),
});

export default connect(
    mapStateToProps,
    null,
);
