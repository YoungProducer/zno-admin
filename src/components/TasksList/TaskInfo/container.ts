// Created by: Oleksandr Bezrukov
// Creation date: 5 February 2020

/**
 * Define main types and interface for 'TaskInfo' component.
 * Export function which connect actions and variables from the store to this component.
 */

// External imports
import { connect } from 'react-redux';

// Application's imports
import {
    deleteTaskAction,
    changeTaskTypeAction,
    changeTaskAnswerAction,
    changeAnswersAmountAction,
    changeTaskImageAction,
    changeExplanationImageAction,
    activateEditionModeAction,
    deactivateEditionModeAction,
} from 'store/slices/createTest';
import {
    deleteTaskImageAction,
    deleteExplanationImageAction,
} from 'store/slices/createTest/tasksList';
import {
    selectTasksListEditionMode,
} from 'store/selectors/createTest';
import { ITask } from 'store/slices/createTest';
import { RootState } from 'store/slices';
import { ETaskType } from 'components/CreateTest/TaskConfigurations/Component';

/**
 * Props which component get from parent component.
 */
interface IOwnProps {
    task: ITask;
    expanded: string | false;
    index: number;
    onChange: (value: string) => any;
}

/**
 * Props which component get from redux store
 */
interface IStateProps {
    editionMode: boolean;
}

/**
 * Props(actions) which component can dispatch
 */
interface IDispatchProps {
    deleteTask: () => void;
    deleteTaskImage: () => void;
    deleteExplanationImage: () => void;
    setTaskAnswer: (payload: number | any[]) => void;
    setAnswersAmount: (payload: number) => void;
    changeTaskType: (payload: ETaskType) => void;
    changeTaskImage: (payload: File) => void;
    changeExplanationImage: (payload: File) => void;
    activateEditionMode: () => void;
    deactivateEditionMode: () => void;
}

/**
 * Define type for 'TaskInfo' component which describe all props pushed to the component.
 */
export type TTaskInfoProps = IOwnProps & IStateProps & IDispatchProps;

/**
 * Define mapStateToProps funciton.
 * This function returns variables from the redux store.
 */
const mapStateToProps = (state: RootState): IStateProps => ({
    editionMode: selectTasksListEditionMode(state),
});

/**
 * Define mapDispatchToProps function.
 * Function returns functions which dispatch actions to store.
 */
const mapDispatchToProps = (dispatch: any, ownProps: IOwnProps): IDispatchProps => ({
    deleteTask: () =>
        dispatch(deleteTaskAction(ownProps.task.id)),

    deleteTaskImage: () =>
        dispatch(deleteTaskImageAction(ownProps.task.id)),

    deleteExplanationImage: () =>
        dispatch(deleteExplanationImageAction(ownProps.task.id)),

    setTaskAnswer: (payload: number | any[]) =>
        dispatch(changeTaskAnswerAction({
            id: ownProps.task.id,
            answer: payload,
        })),

    setAnswersAmount: (payload: number) =>
        dispatch(changeAnswersAmountAction({
            id: ownProps.task.id,
            answersAmount: payload,
        })),

    changeTaskType: (payload: ETaskType) =>
        dispatch(changeTaskTypeAction({
            id: ownProps.task.id,
            type: payload,
        })),

    changeTaskImage: (payload: File) =>
        dispatch(changeTaskImageAction({
            id: ownProps.task.id,
            image: payload,
        })),

    changeExplanationImage: (payload: File) =>
        dispatch(changeExplanationImageAction({
            id: ownProps.task.id,
            image: payload,
        })),

    activateEditionMode: () =>
        dispatch(activateEditionModeAction()),

    deactivateEditionMode: () =>
        dispatch(deactivateEditionModeAction()),
});

/**
 * Export function which connect actions and variables from the redux store to component.
 */
export default connect<IStateProps, IDispatchProps, IOwnProps>(
    mapStateToProps,
    mapDispatchToProps,
);
