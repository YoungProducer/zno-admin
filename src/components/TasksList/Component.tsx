// Created by: Oleksandr Bezrukov
// Creation date: 3 February

/**
 * Component which show current list of tasks.
 * Allow to delete and edit them.
 */

// External imports
import React, { useState, useEffect } from 'react';
import Button from '@material-ui/core/Button';
import Divider from '@material-ui/core/Divider';
import Box from '@material-ui/core/Box';
import Typography from '@material-ui/core/Typography';
import Dialog from '@material-ui/core/Dialog';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import MuiExpansionPanel from '@material-ui/core/ExpansionPanel';
import MuiExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import MuiExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import Collapse from '@material-ui/core/Collapse';
import Slide from '@material-ui/core/Slide';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import { TransitionProps } from '@material-ui/core/transitions';
import { makeStyles, createStyles, Theme, withStyles } from '@material-ui/core/styles';

// Application's imports
import TaskInfo from './TaskInfo';
import { TTasksListProps } from './container';

// Define classes as hook
const useStyles = makeStyles((theme: Theme) => createStyles({
    root: {
        backgroundColor: '#fff',
        borderRadius: 5,
        border: `2px solid #eee`,
        width: `100%`,
        padding: theme.spacing(1),
        marginBottom: theme.spacing(1),
    },
    appBar: {
        position: 'relative',
    },
    title: {
        marginLeft: theme.spacing(2),
        flex: 1,
    },
    innerWrapper: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    icon: {
        height: 48,
    },
}));

// tslint:disable-next-line: ter-prefer-arrow-callback
const Transition = React.forwardRef<unknown, TransitionProps>(function Transition(props, ref) {
    return <Slide direction="up" ref={ref} {...props} />;
});

const Component = ({
    tasksList,
    deleteTask,
}: TTasksListProps) => {
    // Declare and define classes variable
    const classes = useStyles({});

    const [tasksAmountString, setTasksAmountString] = useState<string>('');

    const [expanded, setExpanded] = React.useState<string | false>(false);

    const handleChange = (panel: string) => (event: React.ChangeEvent<{}>, newExpanded: boolean) => {
        setExpanded(newExpanded ? panel : false);
    };

    const [openList, toggleOpenList] = useState<boolean>(false);
    const handleOpenList = () => toggleOpenList(true);
    const handleCloseList = () => toggleOpenList(false);
    const handleToggleList = () => openList ? handleCloseList() : handleOpenList();

    useEffect(() => {
        let lastWord = '';

        if (tasksList.length <= 4) lastWord = 'завдання';
        if (tasksList.length > 4 && tasksList.length <= 20) lastWord = 'завдань';
        if (tasksList.length > 20 && tasksList.length % 10 === 1 || 2 || 3 || 4) lastWord = 'завдання';
        else lastWord = 'завдань';

        setTasksAmountString(`Створено ${tasksList.length} ${lastWord}`);
    },        [tasksList]);

    return (
        <>
            <Collapse
                in={tasksList.length !== 0}
                className={classes.root}
                mountOnEnter
                unmountOnExit
            >
                <Box className={classes.innerWrapper}>
                    <Typography
                        variant='h6'
                        color='primary'
                    >
                        {tasksAmountString}
                    </Typography>
                    <Button
                        color='primary'
                        variant='contained'
                        disableElevation
                        onClick={handleOpenList}
                    >
                        Переглянути
                    </Button>
                </Box>
            </Collapse>
            <Dialog
                fullScreen
                open={openList}
                TransitionComponent={Transition}
            >
                <AppBar className={classes.appBar}>
                    <Toolbar>
                        <IconButton edge="start" color="inherit" onClick={handleCloseList} aria-label="close">
                            <CloseIcon />
                        </IconButton>
                        <Typography variant="h6" className={classes.title}>
                            Список створених завдань
                        </Typography>
                    </Toolbar>
                </AppBar>
                <List>
                    {tasksList.map((task, index) => (
                        <TaskInfo
                            key={index}
                            task={task}
                            index={index}
                            expanded={expanded}
                            onChange={handleChange}
                        />
                    ))}
                </List>
            </Dialog>
        </>
    );
};

export default Component;
