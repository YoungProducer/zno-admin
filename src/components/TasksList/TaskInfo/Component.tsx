// Created by: Oleksandr Bezrukov
// Creation date: 4 February 2020

/**
 * Display info about current task(images, answers, type).
 * Allow delete task or edit it.
 */

// External imports
import React, { useState, useMemo, useCallback } from 'react';
import MuiButton from '@material-ui/core/Button';
import MuiExpansionPanel from '@material-ui/core/ExpansionPanel';
import MuiExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import MuiExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import MuiExpansionPanelActions from '@material-ui/core/ExpansionPanelActions';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import { red } from '@material-ui/core/colors';
import {
    withStyles,
    makeStyles,
    createStyles,
    Theme,
    ThemeProvider,
    createMuiTheme,
} from '@material-ui/core/styles';

// Application's imports
import OneRightAnswer from 'components/CreateTest/TaskConfigurations/AnswerSelection/OneRightAnswer';
import RelationAnswer from 'components/CreateTest/TaskConfigurations/AnswerSelection/RelationAnswer';
import TextAnswer from 'components/CreateTest/TaskConfigurations/AnswerSelection/TextAnswer';
import ImageUploadModal, { TUploadImageType } from 'modals/ImageUploadModal';
import { TypeSelector } from 'components/CreateTest/TaskConfigurations/Component';
import { TTaskInfoProps } from './container';
import { TaskType, TaskSlice, ImageType } from 'store/slices/task';

const expPanelTheme = createMuiTheme({
    palette: {
        primary: {
            main: red[300],
            contrastText: '#fff',
        },
    },
});

// Define classes as hook
const useStyles = makeStyles((theme: Theme) => createStyles({
    grid: {
        flexDirection: 'column',
        justifyContent: 'space-between',
        [theme.breakpoints.up('lg')]: {
            flexDirection: 'row',
        },
    },
    img: {
        width: '90vw',
        [theme.breakpoints.up('lg')]: {
            width: '50vw',
        },
    },
    imageWrapper: {
        borderRadius: 5,
        border: `2px solid #eee`,
        padding: theme.spacing(1),
        marginBottom: theme.spacing(1),
        '&:last-child': {
            marginBottom: 0,
        },
    },
    imageActions: {
        padding: theme.spacing(1),
        '& button': {
            marginLeft: 0,
            '&:not(:first-child)': {
                marginLeft: theme.spacing(1),
            },
        },
    },
}));

const DeclineButton = withStyles(theme => ({
    root: {
        backgroundColor: red[500],
        color: theme.palette.getContrastText(red[500]),
        '&:hover': {
            backgroundColor: red[600],
        },
    },
}))(MuiButton);

const ExpansionPanel = withStyles({
    root: {
        border: '1px solid rgba(0, 0, 0, .125)',
        boxShadow: 'none',
        '&:not(:last-child)': {
            borderBottom: 0,
        },
        '&:before': {
            display: 'none',
        },
        '&$expanded': {
            margin: 'auto',
        },
    },
    expanded: {},
})(MuiExpansionPanel);

const ExpansionPanelSummary = withStyles({
    root: {
        backgroundColor: 'rgba(0, 0, 0, .03)',
        borderBottom: '1px solid rgba(0, 0, 0, .125)',
        marginBottom: -1,
        minHeight: 56,
        '&$expanded': {
            minHeight: 56,
        },
    },
    content: {
        '&$expanded': {
            margin: '12px 0',
        },
    },
    expanded: {},
})(MuiExpansionPanelSummary);

const ExpansionPanelDetails = withStyles(theme => ({
    root: {
        padding: theme.spacing(2),
    },
}))(MuiExpansionPanelDetails);

const ExpansionPanelActions = withStyles(theme => ({
    root: {
        padding: theme.spacing(2),
    },
}))(MuiExpansionPanelActions);

const useUploadImageFields = (props: TTaskInfoProps) => {
    const { updateTask, task } = props;

    const [imageType, setImageType] = useState<TUploadImageType>('task');

    const handleDeleteImage = (type: ImageType) =>
        updateTask({ images: { [type]: null } });

    const handleChangeImage = (type: ImageType, image: File) =>
        updateTask({ images: { [type]: image } });

    const imagePreview = useMemo(() =>
        task.images[imageType] !== null
            ? task.images[imageType].preview
            : '',
        [imageType, task]);

    const setImage = useCallback((image: File) => handleChangeImage(imageType, image),
        [imageType]);

    const deleteImage = useCallback(() => handleDeleteImage(imageType),
        [imageType]);

    const onDropCallback = useCallback((acceptedFiles: File[]) =>
        setImage(acceptedFiles[0]), [imageType]);

    return {
        handleDeleteImage,
        setImageType,
        uploadModalFields: {
            imagePreview,
            setImage,
            deleteImage,
            onDropCallback,
        },
    };
};

const Component = (props: TTaskInfoProps) => {
    // Declare and define classes variable
    const classes = useStyles({});

    const {
        index,
        updateTask,
        deleteTask,
        task,
        expanded,
        deactivateEditionMode,
        activateEditionMode,
        editionMode,
        onChange,
    } = props;

    const {
        uploadModalFields,
        handleDeleteImage,
        setImageType,
    } = useUploadImageFields(props);

    const [modalOpen, toggleModalOpen] = useState<boolean>(false);
    const handleOpenModal = () => toggleModalOpen(true);
    const handleCloseModal = () => toggleModalOpen(false);

    // Handle change in radio group
    const handleChangeTaskType = (event: React.ChangeEvent<HTMLInputElement>) => {
        updateTask({
            type: (event.target as HTMLInputElement).value as TaskType,
        });
    };

    const handleChangeAnswersAmount = (amount: number) =>
        updateTask({ answersAmount: amount });

    const handleChangeAnswer = (payload: TaskSlice.SetAnswerPayload) =>
        updateTask({ answer: payload });

    const {
        type,
        answer,
        answersAmount,
        images,
    } = task;

    const handleClickEditButton = () => {
        if (editionMode) {
            deactivateEditionMode();
        } else {
            activateEditionMode();
        }
    };

    return (
        <>
            <ExpansionPanel
                component='li'
                square
                expanded={expanded === `panel${task.id}`}
                onChange={onChange(`panel${task.id}`)}
            >
                <ExpansionPanelSummary
                    aria-controls={`panel${task.id}d-content`}
                    id={`panel${task.id}d-header`}
                >
                    <Typography
                        data-testid="task-info-panel-summary-text"
                        // color={error ? 'error' : 'initial'}
                    >
                        {`Завдання №${index + 1}`}
                    </Typography>
                </ExpansionPanelSummary>
                <ExpansionPanelDetails>
                    <Grid
                        container
                        className={classes.grid}
                        spacing={2}
                    >
                        <Grid item>
                            <Typography>
                                Тип завдання: {
                                    type === 'SINGLE'
                                    ? 'одна правильна відповідь'
                                        : type === 'RELATIONS'
                                        ? 'відношення'
                                    : 'текстова відповідь'
                                }
                            </Typography>
                            { editionMode && (
                                <TypeSelector
                                    value={type}
                                    onChange={handleChangeTaskType}
                                />
                            )}
                            {type === 'SINGLE' && (
                                <OneRightAnswer
                                    answer={answer}
                                    answersAmount={answersAmount}
                                    setTaskAnswer={handleChangeAnswer}
                                    setAnswersAmount={handleChangeAnswersAmount}
                                />
                            )}
                            {type === 'RELATIONS' && (
                                <RelationAnswer
                                    answer={answer}
                                    answersAmount={answersAmount}
                                    setTaskAnswer={handleChangeAnswer}
                                    setAnswersAmount={handleChangeAnswersAmount}
                                />
                            )}
                            {type === 'TEXT' && (
                                <TextAnswer
                                    answer={answer}
                                    answersAmount={task.answersAmount}
                                    setTaskAnswer={handleChangeAnswer}
                                    setAnswersAmount={handleChangeAnswersAmount}
                                />
                            )}
                        </Grid>
                        <Grid item>
                            <div className={classes.imageWrapper}>
                                <Typography>
                                    Зображення завдання: {!images.task && 'зображення не завантажено.'}
                                </Typography>
                                {
                                    expanded === `panel${task.id}`
                                    && images.task
                                    && <img src={images.task.preview} className={classes.img} />
                                }
                                { editionMode && (
                                    <div className={classes.imageActions}>
                                        <MuiButton
                                            color='primary'
                                            variant='outlined'
                                            onClick={() => {
                                                setImageType('task');
                                                handleOpenModal();
                                            }}
                                        >
                                            Змінити
                                        </MuiButton>
                                        <DeclineButton
                                            variant='contained'
                                            disableElevation
                                            onClick={() => handleDeleteImage('task')}
                                        >
                                            Видалити
                                        </DeclineButton>
                                    </div>
                                )}
                            </div>
                            <div className={classes.imageWrapper}>
                                <Typography>
                                    Зображення пояснення: {!images.explanation && 'зображення не завантажено.'}
                                </Typography>
                                {
                                    expanded === `panel${task.id}`
                                    && images.explanation
                                    && <img src={images.explanation.preview} className={classes.img} />
                                }
                                { editionMode && (
                                    <div className={classes.imageActions}>
                                        <MuiButton
                                            color='primary'
                                            variant='outlined'
                                            onClick={() => {
                                                setImageType('explanation');
                                                handleOpenModal();
                                            }}
                                        >
                                            Змінити
                                        </MuiButton>
                                        <DeclineButton
                                            variant='contained'
                                            disableElevation
                                            onClick={() => handleDeleteImage('explanation')}
                                        >
                                            Видалити
                                        </DeclineButton>
                                    </div>
                                )}
                            </div>
                        </Grid>
                    </Grid>
                </ExpansionPanelDetails>
                <ExpansionPanelActions>
                    <MuiButton
                        color='primary'
                        variant='contained'
                        disableElevation
                        onClick={handleClickEditButton}
                    >
                        { editionMode ? 'Закінчити редагування' : 'Редагувати'}
                    </MuiButton>
                    <DeclineButton
                        variant='contained'
                        disableElevation
                        onClick={deleteTask}
                    >
                        Видалити
                    </DeclineButton>
                </ExpansionPanelActions>
            </ExpansionPanel>
            <ImageUploadModal
                open={modalOpen}
                onClose={handleCloseModal}
                {...uploadModalFields}
            />
        </>
    );
};

export default Component;
