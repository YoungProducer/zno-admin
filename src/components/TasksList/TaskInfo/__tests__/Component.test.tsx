/**
 * Created by: Oleksandr Bezrukov
 * Creation date 17 February
 *
 * Create test suites for TaskInfo component.
 */

// External imports
import React from 'react';
import { shallow } from 'enzyme';

// Application's imports
import Component from '../Component';
import { TTaskInfoProps } from '../container';
import { ETaskType, ITask } from 'store/slices/createTest';

describe('TaskInfo component', () => {
    const requiredProps: TTaskInfoProps = {
        activateEditionMode: jest.fn(),
        deactivateEditionMode: jest.fn(),
        changeExplanationImage: jest.fn(),
        changeTaskImage: jest.fn(),
        changeTaskType: jest.fn(),
        deleteExplanationImage: jest.fn(),
        deleteTask: jest.fn(),
        deleteTaskImage: jest.fn(),
        onChange: jest.fn(),
        setAnswersAmount: jest.fn(),
        setTaskAnswer: jest.fn(),
        editionMode: false,
        expanded: false,
        index: 0,
        task: {
            answer: 0,
            answersAmount: 1,
            explanationImage: null,
            taskImage: null,
            id: 0,
            taskType: ETaskType.ONE_RIGHT,
        },
    };

    test('Is matches snapshot', () => {
        const root = shallow(<Component { ...requiredProps }/>);

        expect(root).toMatchSnapshot();
    });

    test(`If task property 'error' equals false panel summary text must have initial(black) color`, () => {
        // Define task with error: true
        const task: ITask = {
            answer: 0,
            answersAmount: 1,
            id: 0,
            taskImage: null,
            explanationImage: null,
            taskType: ETaskType.ONE_RIGHT,
        };

        // Get shallow rendered component
        const root = shallow(<Component { ...requiredProps } task={task}/>);

        // Check is typography exists
        expect(root.find(`[data-testid="task-info-panel-summary-text"]`).exists()).toEqual(true);

        // Check is typegraphy has color 'error'
        expect(root.find(`[data-testid="task-info-panel-summary-text"]`).props().color).toEqual('initial');
    });

    test(`If task property 'error' equals true panel summary text must have error(red) color`, () => {
        // Define task with error: true
        const task: ITask = {
            answer: 0,
            answersAmount: 1,
            id: 0,
            taskImage: null,
            explanationImage: null,
            taskType: ETaskType.ONE_RIGHT,
            error: true,
        };

        // Get shallow rendered component
        const root = shallow(<Component { ...requiredProps } task={task}/>);

        // Check is typography exists
        expect(root.find(`[data-testid="task-info-panel-summary-text"]`).exists()).toEqual(true);

        // Check is typegraphy has color 'error'
        expect(root.find(`[data-testid="task-info-panel-summary-text"]`).props().color).toEqual('error');
    });
});
