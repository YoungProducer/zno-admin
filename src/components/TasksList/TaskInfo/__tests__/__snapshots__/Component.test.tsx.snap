// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskInfo component Is matches snapshot 1`] = `
<Fragment>
  <WithStyles(WithStyles(ForwardRef(ExpansionPanel)))
    component="li"
    expanded={false}
    square={true}
  >
    <WithStyles(WithStyles(ForwardRef(ExpansionPanelSummary)))
      aria-controls="panel0d-content"
      id="panel0d-header"
    >
      <WithStyles(ForwardRef(Typography))
        color="initial"
        data-testid="task-info-panel-summary-text"
      >
        Завдання №1
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(WithStyles(ForwardRef(ExpansionPanelSummary)))>
    <WithStyles(WithStyles(ForwardRef(ExpansionPanelDetails)))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-grid-1"
        container={true}
        spacing={2}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Typography))>
            Тип завдання: 
            одна правильна відповідь
          </WithStyles(ForwardRef(Typography))>
          <Component
            answer={0}
            answersAmount={1}
            setAnswersAmount={[MockFunction]}
            setTaskAnswer={[MockFunction]}
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <div
            className="makeStyles-imageWrapper-3"
          >
            <WithStyles(ForwardRef(Typography))>
              Зображення завдання: 
              зображення не завантажено.
            </WithStyles(ForwardRef(Typography))>
          </div>
          <div
            className="makeStyles-imageWrapper-3"
          >
            <WithStyles(ForwardRef(Typography))>
              Зображення пояснення: 
              зображення не завантажено.
            </WithStyles(ForwardRef(Typography))>
          </div>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(WithStyles(ForwardRef(ExpansionPanelDetails)))>
    <WithStyles(WithStyles(ForwardRef(ExpansionPanelActions)))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disableElevation={true}
        onClick={[Function]}
        variant="contained"
      >
        Редагувати
      </WithStyles(ForwardRef(Button))>
      <WithStyles(WithStyles(ForwardRef(Button)))
        disableElevation={true}
        onClick={[MockFunction]}
        variant="contained"
      >
        Видалити
      </WithStyles(WithStyles(ForwardRef(Button)))>
    </WithStyles(WithStyles(ForwardRef(ExpansionPanelActions)))>
  </WithStyles(WithStyles(ForwardRef(ExpansionPanel)))>
  <Connect(Component)
    deleteExplanationImage={[MockFunction]}
    deleteTaskImage={[MockFunction]}
    explanationImage={null}
    onClose={[Function]}
    open={false}
    setExplanationImage={[MockFunction]}
    setTaskImage={[MockFunction]}
    taskImage={null}
    uploadImageType="task"
  />
</Fragment>
`;
